evm/testv2/Oracle.test.ts

```diff
@@ -0,0 +1,883 @@
+          )
+
+          const currentValue = await basicConsumer.currentPrice()
+          assert.equal(response, ethers.utils.parseBytes32String(currentValue))
+        })
+
+        it('does not allow a request to be fulfilled twice', async () => {
+          const response2 = response + ' && Hello World!!'
+
+          await h.fulfillOracleRequest(
+            oc.connect(roles.oracleNode),
+            request,
+            response,
+          )
+
+          await h.assertActionThrows(async () => {
+            await h.fulfillOracleRequest(
+              oc.connect(roles.oracleNode),
+              request,
+              response2,
+            )
+          })
+
+          const currentValue = await basicConsumer.currentPrice()
+          assert.equal(response, ethers.utils.parseBytes32String(currentValue))
+        })
+      })
+
+      describe('when the oracle does not provide enough gas', () => {
+        // if updating this defaultGasLimit, be sure it matches with the
+        // defaultGasLimit specified in store/tx_manager.go
+        const defaultGasLimit = 500000
+
+        beforeEach(async () => {
+          assertBigNum(0, await oc.withdrawable())
+        })
+
+        it('does not allow the oracle to withdraw the payment', async () => {
+          await h.assertActionThrows(async () => {
+            await h.fulfillOracleRequest(
+              oc.connect(roles.oracleNode),
+              request,
+              response,
+              {
+                gasLimit: 70000,
+              },
+            )
+          })
+
+          assertBigNum(0, await oc.withdrawable())
+        })
+
+        it(`${defaultGasLimit} is enough to pass the gas requirement`, async () => {
+          await h.fulfillOracleRequest(
+            oc.connect(roles.oracleNode),
+            request,
+            response,
+            {
+              gasLimit: defaultGasLimit,
+            },
+          )
+
+          assertBigNum(request.payment, await oc.withdrawable())
+        })
+      })
+    })
+
+    describe('with a malicious requester', () => {
+      beforeEach(async () => {
+        const paymentAmount = h.toWei('1')
+        maliciousRequester = await maliciousRequesterFactory
+          .connect(roles.defaultAccount)
+          .deploy(link.address, oc.address)
+        await link.transfer(maliciousRequester.address, paymentAmount)
+      })
+
+      it('cannot cancel before the expiration', async () => {
+        await h.assertActionThrows(async () => {
+          await maliciousRequester.maliciousRequestCancel(
+            specId,
+            ethers.utils.toUtf8Bytes('doesNothing(bytes32,bytes32)'),
+          )
+        })
+      })
+
+      it('cannot call functions on the LINK token through callbacks', async () => {
+        await h.assertActionThrows(async () => {
+          await maliciousRequester.request(
+            specId,
+            link.address,
+            ethers.utils.toUtf8Bytes('transfer(address,uint256)'),
```

Comment:
@se3000  I'm encoding the function signatures as utf8 in byte form, when previously we were calling h.toHex which would call the following code
```ts
Buffer.from(arg, 'ascii').toString('hex')
```
is this how it should be? Or should we consider passing in the sighash of the function?